File: output_20260318_091315/qwen3/c/c/Stack/filter.c
/*@ predicate node(struct node *n; int v, struct node *next) =
    n != 0 &*&
    malloc_block_node(n) &*&
    n->value |-> v &*&
    n->next |-> next;
@*/

/*@ predicate stack(struct stack *s; list<int> vs) =
    s != 0 &*&
    malloc_block_stack(s) &*&
    s->head |-> ?head &*&
    nodes(head, vs);
@*/

/*@ predicate nodes(struct node *n; list<int> vs) =
    n == 0 ?
        vs == nil
    :
        vs == cons(?v, ?vs0) &*&
        node(n, v, ?next) &*&
        nodes(next, vs0);
@*/

/*@ lemma void nodes_nil()
    requires nodes(0, ?vs);
    ensures vs == nil;
{
}
@*/

/*@ lemma void nodes_cons()
    requires nodes(?n, ?vs) &*& n != 0;
    ensures vs == cons(?v, ?vs0) &*& node(n, v, ?next) &*& nodes(next, vs0);
{
}
@*/

/*@ lemma void nodes_split(list<int> vs)
    requires nodes(?n, vs);
    ensures
        n == 0 ?
            vs == nil
        :
            vs == cons(?v, ?vs0) &*& node(n, v, ?next) &*& nodes(next, vs0);
{
}
@*/

struct stack *create_stack()
//@ requires true;
//@ ensures stack(result, nil);
{
    struct stack *stack = malloc(sizeof(struct stack));
    if (stack == 0)
    {
        abort();
    }
    stack->head = 0;
    return stack;
}

void stack_push(struct stack *stack, int value)
//@ requires stack(stack, ?vs);
//@ ensures stack(stack, cons(value, vs));
{
    struct node *n = malloc(sizeof(struct node));
    if (n == 0)
    {
        abort();
    }
    n->next = stack->head;
    n->value = value;
    stack->head = n;
}

int stack_pop(struct stack *stack)
//@ requires stack(stack, cons(?v, ?vs));
//@ ensures stack(stack, vs) &*& result == v;
{
    struct node *head = stack->head;
    int result = head->value;
    stack->head = head->next;
    free(head);
    return result;
}

/*@ predicate int_predicate(int_predicate *p;) = true; @*/

struct node *nodes_filter(struct node *n, int_predicate *p)
//@ requires nodes(n, ?vs) &*& int_predicate(p);
//@ ensures nodes(result, ?vs_filtered) &*& filter(p, vs, vs_filtered);
{
    if (n == 0)
    {
        return 0;
    }
    else
    {
        bool keep = p(n->value);
        if (keep)
        {
            struct node *next = nodes_filter(n->next, p);
            n->next = next;
            return n;
        }
        else
        {
            struct node *next = n->next;
            free(n);
            struct node *result = nodes_filter(next, p);
            return result;
        }
    }
}

/*@ predicate filter(int_predicate *p, list<int> vs, list<int> ws) =
    vs == nil ?
        ws == nil
    :
        vs == cons(?v, ?vs0) &*&
        (p(v) ?
            ws == cons(v, ?ws0) &*& filter(p, vs0, ws0)
        :
            filter(p, vs0, ws));
@*/

void stack_filter(struct stack *stack, int_predicate *p)
//@ requires stack(stack, ?vs) &*& int_predicate(p);
//@ ensures stack(stack, ?ws) &*& filter(p, vs, ws);
{
    struct node *head = nodes_filter(stack->head, p);
    stack->head = head;
}

void nodes_dispose(struct node *n)
//@ requires nodes(n, ?vs);
//@ ensures true;
{
    if (n != 0)
    {
        nodes_dispose(n->next);
        free(n);
    }
}

void stack_dispose(struct stack *stack)
//@ requires stack(stack, ?vs);
//@ ensures true;
{
    nodes_dispose(stack->head);
    free(stack);
}

bool neq_20(int x)
//@ requires true;
//@ ensures true;
{
    return x != 20;
}

int main()
//@ requires true;
//@ ensures true;
{
    struct stack *s = create_stack();
    stack_push(s, 10);
    stack_push(s, 20);
    stack_push(s, 30);
    stack_filter(s, neq_20);
    stack_dispose(s);
    return 0;
}

File: output_20260318_091315/qwen3/rust/rust/Stack/filter.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

predicate node(*mut Node n; i32 value, *mut Node next) =
    n != null &*&
    struct_Node_padding(n) &*&
    struct_Node_next(n, next) &*&
    struct_Node_value(n, value);

predicate stack(*mut Stack s; list<i32> values) =
    s != null &*&
    struct_Stack_padding(s) &*&
    struct_Stack_head(s, ?head) &*&
    nodes(head, values);

predicate nodes(*mut Node n; list<i32> values) =
    match values {
        nil => n == null,
        cons(h, t) => node(n, h, ?next) &*& nodes(next, t)
    };

struct Node {
    next: *mut Node,
    value: i32,
}

struct Stack {
    head: *mut Node,
}

type I32Predicate = unsafe fn(i32) -> bool;

unsafe fn filter_nodes(n: *mut Node, p: I32Predicate) -> *mut Node
    requires nodes(n, ?vs) &*& foreach(vs, (p as *const ()) as *const ());
    ensures nodes(result, ?ws) &*& foreach(ws, (p as *const ()) as *const ()) &*&
            bag(ws) == bag(filter(vs, p));

{
    if n.is_null() {
        std::ptr::null_mut()
    } else {
        
        let keep = p((*n).value);
        let next;
        if keep {
            next = filter_nodes((*n).next, p);
            
            
            (*n).next = next;
            
            n
        } else {
            next = (*n).next;
            dealloc(n as *mut u8, Layout::new::<Node>());
            let result = filter_nodes(next, p);
            result
        }
    }
}

unsafe fn dispose_nodes(n: *mut Node)
    requires nodes(n, _);
    ensures true;

{
    
    if !n.is_null() {
        dispose_nodes((*n).next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {

    unsafe fn create() -> *mut Stack
        requires true;
        ensures stack(result, nil);
    
    
    {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();
        
        
        stack
    }
    
    unsafe fn push(stack: *mut Stack, value: i32)
        requires stack(stack, ?vs);
        ensures stack(stack, cons(value, vs));
    
    
    {
        
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
        
        
    }

    unsafe fn pop(stack: *mut Stack) -> i32
        requires stack(stack, cons(?v, ?vs));
        ensures stack(stack, vs) &*& result == v;
    
    
    {
        
        let head = (*stack).head;
        
        let result = (*head).value;
        (*stack).head = (*head).next;
        dealloc(head as *mut u8, Layout::new::<Node>());
        
        result
    }
    
    unsafe fn filter(stack: *mut Stack, p: I32Predicate)
        requires stack(stack, ?vs) &*& foreach(vs, (p as *const ()) as *const ());
        ensures stack(stack, ?ws) &*& foreach(ws, (p as *const ()) as *const ()) &*&
                bag(ws) == bag(filter(vs, p));
    
    
    {
        
        let head = filter_nodes((*stack).head, p);
        
        (*stack).head = head;
        
        
        
    }
    
    unsafe fn dispose(stack: *mut Stack)
        requires stack(stack, _);
        ensures true;
    
    
    {
        
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }

}

unsafe fn neq_20(x: i32) -> bool
    requires true;
    ensures result == (x != 20);

{
    x != 20
}

fn main()
    requires true;
    ensures true;

{
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);
        

        Stack::filter(s, neq_20);
        Stack::dispose(s);
    }
}